src/replica/HttpAsyncClientApp.cc

```diff
@@ -0,0 +1,204 @@
+        " assumed for the request.",
+        _expirationIvalSec
+    ).option(
+        "file",
+        "A path to an output file where the response body received from a remote source will"
+        "  be written. This option is ignored if the flag --body is not specified.",
+        _file
+    ).flag(
+        "result2json",
+        "If specified the flag will cause the application to interpret the response body as"
+        " a JSON object.",
+        _result2json
+    ).flag(
+        "verbose",
+        "The flag that allows printing the completion status and the response header"
+        " info onto the standard output stream.",
+        _verbose
+    ).flag(
+        "body",
+        "The flag that allows printing the complete response body. If the --file=<path> option"
+        " is specified then the body will be written into that files. Otherwise it will be"
+        " printed onto the standard output stream.",
+        _body
+    );
+}
+
+
+int HttpAsyncClientApp::runImpl() {
+    string const context = "HttpAsyncClientApp::" + string(__func__) + "  ";
+    const char* allowedMethods[] = {"GET", "POST", "PUT", "DELETE"};
+    if (count(cbegin(allowedMethods), cend(allowedMethods), _method) == 0) {
+        throw invalid_argument(context + "unknown HTTP method: " + _method);
+    }
+    map<string, string> headers;
+    if (!_header.empty()) {
+        auto const pos = _header.find(":");
+        if (pos == string::npos) {
+            headers[_header] = string();
+        } else {
+            headers[_header.substr(0, pos)] = _header.substr(pos + 1);
+        }
+    }
+    ostream* osPtr = nullptr;
+    ofstream fs;
+    if (_body) {
+        if (_file.empty()) {
+            osPtr = &cout;
+        } else {
+            fs.open(_file, ios::out|ios::trunc|ios::binary);
+            if (!fs.is_open()) {
+                throw runtime_error(context + "failed to open/create file: " + _file);
+            }
+            osPtr = &fs;
+        }
+    }
+    boost::asio::io_service io_service;
+    auto const ptr = HttpAsyncClient::create(io_service,
+        [this, osPtr] (auto const& ptr) {
+            this->_dump(ptr, osPtr);
+        },
+        _method, _url, _data, headers, _maxResponseBodySize, _expirationIvalSec);
+
+    ptr->start();
+    io_service.run();
+
+    if (nullptr != osPtr) {
+        osPtr->flush();
+        if (fs.is_open()) fs.close();
+    }
+    return ptr->state() == HttpAsyncClient::State::FINISHED ? 0 : 1;
+}
+
+
+void HttpAsyncClientApp::_dump(shared_ptr<HttpAsyncClient> const& ptr, ostream* osPtr) const {
+    if (_verbose) {
+        cout << "Request completion state: " << HttpAsyncClient::state2str(ptr->state())
+                << ", error: " << ptr->error() << endl;
+    }
+    set<HttpAsyncClient::State> const headerAvailableStates = {
+        HttpAsyncClient::State::FINISHED,
+        HttpAsyncClient::State::BODY_LIMIT_ERROR
+    };
+    if (headerAvailableStates.count(ptr->state()) != 0) {
```

Comment:
This is a very convoluted and inefficient way to write

    if (ptr->state() == HttpAsyncClient::State::FINISHED or ptr->state() == HttpAsyncClient::State::BODY_LIMIT_ERROR) {

Reply:
May be.